### ISSUE TITLE
Ability to provide getFn for specific paths

### ISSUE THREAD

**Issue by cdutson** [krisk/Fuse#414]
So I'm working with a dataset that has localized strings. that is to say that it has multiple objects that contain the string to search against. Example
```
{
    id: 1,
    localizations: [
        {
            culture: 'en-CA',
            name: 'hello'
        },
        {
            culture: 'fr-CA',
            name: 'bonjour'
        },
    ]
}
```
Then, I set the search path for fuse to `localizations.name`

So right now, if I perform a search against this data set with the value of "hello", I'll hit that first item. and if i search "bonjour" I'll hit that first item. This is fine _except_ in my applications case I am localized to a specific culture. So if I'm running the app in 'en-CA', I wouldn't want the fr-CA to ever get hit.

As a work-around, I could simply strip the data out of the collection prior to searching, but that's an intensive action, and I'm not keep on destroying a data structure just to search.

I've worked around this presently by providing a custom getFn:
```
getFn: (item, path) =>
    {
        if (path === 'localizations.name')
        {
            // my custom function for getting the correct localization based on culture
            return getLocalization(item.localizations, "current culture");
        }
        return deepGet(item, path);
},
```

Note: deepGet is a copy/paste of the get function in fuse/helpers. As an aside it would be great to expose that function for exactly this sort of case.

My ask: be able to provide a getFn based on the key. Something akin to:
```
getFns: {
    'localizations.name`: (item, path) => { /* perform specific code for this path only */}
}
```

Basically saying "if there's a key for this path, use this specific getFn, else, use the built-in function.

**Comment by krisk** [comment]
Will think about this, as it has been proposed a couple of time already. 

As for this:

> Note: deepGet is a copy/paste of the get function in fuse/helpers. As an aside it would be great to expose that function for exactly this sort of case.

You can already do this, via `Fuse.config`. See the [documentation](https://fusejs.io/api/config.html).

**Comment by mjbcopland** [comment]
You can also do this by providing a function as the key and having `getFn` call it. I don't know if this is supported (I suspect not) but it provides a very flexible way of achieving what you want.

I'm doing something similar; the items I want to search through are not plain objects and have more complex methods of accessing values.

```ts
interface Person {
  getName(): string
}

declare const people: Person[];
```

If I want to search on names, I would like to do something like this

```ts
new Fuse(people, { keys: [(person: Person) => person.getName()] }).search("John Smith")
```

This actually works if I provide a custom `getFn` but fails type checking because Fuse declares that keys are strings. If you're not using TypeScript then this should work for you, but given it isn't documented, it probably isn't supported and could break in future versions.

```ts
function getFn<T>(item: T, key: string | ((item: T) => string)) {
  return typeof key === 'function' ? key(item) : Fuse.config.getFn(item, key);
}
```

imo this is slightly better than your `getFn` because it is not dependent on the implementation details of `keys` or `items` so could reasonably be included as part of the library as the default behaviour.

```ts
const keys = [{ name: (person: Person) => person.getName() }];

// @ts-ignore keys are expected to be strings
const results = new Fuse(people, { keys, getFn }).search("John Smith");
```

Note that you also need to use the object form for the key, i.e. `[{ name: key }]` and not `[key]`, because functions have a `name` property so [this line](https://github.com/krisk/Fuse/blob/9c80b82/src/tools/KeyStore.js#L23) will use the function's name and not the function itself.

As you mentioned, there are other workarounds such as modifying the data before searching, but these add extra overhead and can get ugly, especially as we increase the number of keys and/or the complexity of the items to be searched.

I could probably draft a PR to add support for function keys but it would be good to keep this discussion going as to whether this is something that we'd like to support officially, and to what extent. For example, we could also include something like allowing `{ keys: ["getName"] }` in the example above, i.e. if the path resolves to a function then call it and use its result as the value.

In the meantime, I've mocked up a quick example using your `localizations` data (and some TypeScript suppressions) which you can play around with [here](https://codesandbox.io/s/boring-mountain-sozuk). You can search on the basic key `"id"` or the more complex function key `getLocalization`.
